uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Import/Parser/Base.php

```diff
@@ -563,4 +573,81 @@ private function getAbsenceStatuses() {
 
     return array_merge($oldAbsenceStatusesMap, $absenceStatuses);
   }
+
+  /**
+   * Returns the calculated entitlement for a Contact within
+   * the AbsencePeriod for the AbsenceType.
+   *
+   * @param array $params
+   *
+   * @return \CRM_HRLeaveAndAbsences_BAO_LeavePeriodEntitlement|null
+   */
+  private function getPeriodEntitlementForContact($params) {
+    $startDate = CRM_Utils_Date::formatDate($params['start_date'], $this->dateFormatType);
+    $endDate = CRM_Utils_Date::formatDate($params['end_date'], $this->dateFormatType);
+    $period = AbsencePeriod::getPeriodContainingDates(new DateTime($startDate), new DateTime($endDate));
+
+    $leavePeriodEntitlement = LeavePeriodEntitlement::getPeriodEntitlementForContact(
+      $params['contact_id'],
+      $period->id,
+      $params['type_id']
+    );
+
+    return $leavePeriodEntitlement;
+  }
+
+  /**
+   * This method checks and ensures that the balance change for the leave request to
+   * be created is not greater than the remaining balance of the period if the
+   * Request’s AbsenceType do not allow overuse.
+   * In case the contact does not have a period entitlement for the absence type, an
+   * appropriate exception is thrown.
+   *
+   * @param array $params
+   *
+   * @throws \CRM_HRLeaveAndAbsences_Exception_InvalidLeaveRequestException
+   **/
+  private function validateEntitlementAndBalanceChange($params) {
+    $isToil = $params['request_type'] == LeaveRequest::REQUEST_TYPE_TOIL;
+    $isCancelled = in_array($params['status_id'], LeaveRequest::getCancelledStatuses());
+
+    //TOIL accrual is independent of Current Balance.
+    //Leave Request is able to be rejected or cancelled disregarding the balance
+    if($isToil || $isCancelled) {
+      return;
+    }
+
+    $leavePeriodEntitlement = $this->getPeriodEntitlementForContact($params);
+    if(!$leavePeriodEntitlement) {
+      throw new InvalidLeaveRequestException(
+        'Contact does not have period entitlement for the absence type',
+        'leave_request_contact_has_no_entitlement',
+        'type_id'
+      );
+    }
+
+    $leaveRequestBalance = $params['total_qty'];
+    $currentBalance = $leavePeriodEntitlement->getBalance();
+    $absenceType = AbsenceType::findById($params['type_id']);
+
+    if(!$absenceType->allow_overuse && $leaveRequestBalance > $currentBalance) {
+      throw new InvalidLeaveRequestException(
+        'There are only '. $currentBalance .' days leave available. This request cannot be made or approved',
```

Comment:
This message will be included in the errors report after import, right? If that's true, I think that "This request cannot be made or approved" won't make sense, as I'm not trying to make or approve anything, but to import some data.